Render custom overlay template parts in Navigation block (behind experiment)

getdave opened the pull request:
## What

This PR implements rendering of custom overlay template parts in the Navigation block, addressing a significant part of #73081. When a custom overlay template part is selected, the block now renders both the standard desktop navigation blocks and the overlay template part blocks, with proper show/hide behavior based on breakpoint and menu state.

## Why

This enables users to create custom overlay designs for mobile navigation while maintaining the standard desktop navigation. The overlay template part can contain any blocks, allowing for fully customized mobile menu experiences.

## How

- Added `get_overlay_blocks_from_template_part()` helper to retrieve blocks from overlay template parts
- Modified `get_responsive_container_markup()` to render both desktop and overlay blocks in separate containers
- Added CSS rules to show/hide desktop and overlay containers based on breakpoint and menu state
- Removed default spacing and colors when custom overlay is present
- Added support for `core/navigation-overlay-close` block as alternative close button
- Conditionally show/hide default close button based on presence of overlay close block
- Added interactivity API directives to overlay close block for closing functionality
- Disabled overlay menu for navigation blocks within overlay template parts to prevent nested overlays
- Added spacing to default close button when rendered within custom overlay

## Markup Changes and Backwards Compatibility

**Important**: The desktop container wrapper (`wp-block-navigation__desktop-container`) is gated behind the `gutenberg-customizable-navigation-overlays` experiment to ensure backward compatibility.

- **With experiment enabled**: Desktop navigation is wrapped in `<div class="wp-block-navigation__desktop-container">` to enable proper show/hide behavior with overlay containers
- **Without experiment enabled**: Desktop navigation markup remains exactly as before (no wrapper div), maintaining full backward compatibility
- This ensures that users without the experiment enabled will not experience any CSS or layout changes to their existing navigation blocks

The experiment gate prevents the markup change from affecting trunk until the feature is ready for general availability. All other functionality (overlay rendering, close button handling, etc.) only activates when an overlay is selected, which the UI prevents unless the experiment is enabled.

## Default/Fallback Scenarios

- When no overlay is selected: Navigation block behaves exactly as before (backward compatible)
- When overlay template part is not found: Falls back to desktop-only rendering
- When overlay has no blocks: Falls back to desktop-only rendering
- Default close button is shown when overlay doesn't contain `core/navigation-overlay-close` block

## Testing Instructions

1. **Overlays containing Nav blocks with overlay set to "Always"**:
   - Create an overlay template part that contains a Navigation block
   - Set the Navigation block's overlay menu to "Always"
   - Verify that the nested Navigation block does NOT render its own overlay (should be disabled to prevent inception)
   - The nested Navigation block should render as a standard navigation without overlay functionality

2. **Overlays with/without Nav Overlay Close block**:
   - **With `core/navigation-overlay-close` block**: Create an overlay template part that includes a `core/navigation-overlay-close` block. Verify that:
     - The default close button is hidden
     - The custom overlay close block can close the overlay when clicked
   - **Without `core/navigation-overlay-close` block**: Create an overlay template part without the close block. Verify that:
     - The default close button is visible and positioned with spacing from edges
     - The default close button can close the overlay when clicked

3. **With and without custom overlay**:
   - **Without custom overlay**: Verify that the Navigation block behaves exactly as before (backward compatible)
     - Desktop navigation displays normally
     - Mobile overlay works with default styling
     - Default close button appears and functions correctly
   - **With custom overlay**: Verify that:
     - Desktop navigation displays normally (desktop container visible)
     - Mobile overlay shows custom overlay content when menu is opened (overlay container visible)
     - Custom overlay content is properly styled (no default padding/colors applied)
     - Both containers are properly hidden/shown based on breakpoint

## Accessibility and SEO Considerations

⚠️ **Important Note**: This implementation renders overlay content directly in the DOM, which may lead to accessibility and SEO concerns:

- **Accessibility**: Overlay content (including any navigation blocks within it) is present in the DOM even when visually hidden. While we use `aria-hidden="true"` by default, this may not be sufficient for all assistive technologies.

- **SEO**: Search engines may index overlay content that is intended to be mobile-only, potentially causing duplicate content issues if the overlay contains navigation blocks.

### Potential Mitigations

To address these concerns, we could consider:

1. **Dynamic `aria-hidden` management**: Use Interactivity API to toggle `aria-hidden` based on breakpoint and menu state (currently only set statically)
2. **`hidden` attribute**: Add `hidden` attribute when overlay is not visible (in addition to CSS display:none)
3. **`inert` attribute**: Use `inert` attribute to make overlay content non-interactive when hidden
4. **Meta tags**: Add `data-noindex` or similar attributes to overlay container when hidden
5. **Lazy loading**: Consider lazy loading overlay content only when menu is opened (future enhancement)

These mitigations should be evaluated and potentially implemented in a follow-up PR based on testing and feedback.

getdave committed:
Render custom overlay template parts in Navigation block

- Add get_overlay_blocks_from_template_part() helper to retrieve blocks from overlay template parts
- Add get_template_part_blocks_html() to render template part blocks without navigation container wrapper
- Modify get_responsive_container_markup() to render both desktop and overlay blocks in separate containers
- Add CSS rules to show/hide desktop and overlay containers based on breakpoint and menu state
- Ensure overlay container is full width
- Maintain backward compatibility when no overlay is selected

getdave committed:
Remove default spacing and colors when custom overlay is rendered

- Add has-custom-overlay class to responsive container when overlay template part is selected
- Exclude default padding from responsive-container when custom overlay is present
- Exclude default padding-top from responsive-container-content when custom overlay is present
- Exclude default margin-top from responsive-dialog when custom overlay is present
- Exclude default background and text colors when custom overlay is present
- Simplify CSS structure by removing duplicate selectors

getdave committed:
Conditionally show/hide default close button based on overlay close block

- Add has_navigation_overlay_close_block() helper to check for core/navigation-overlay-close block
- Skip checking inner blocks of navigation blocks for optimization
- Conditionally render default close button only when overlay doesn't have its own close block
- Hide default close button when core/navigation-overlay-close block is present in overlay template part

getdave committed:
Add interactivity API directives to navigation-overlay-close block

- Add block_core_navigation_add_directives_to_overlay_close() function following the same pattern as block_core_navigation_add_directives_to_submenu()
- Use WP_HTML_Tag_Processor to find and add directives to the overlay close button
- Add data-wp-on--click directive to enable closing the overlay menu
- Apply directives only when overlay close block is present and navigation is interactive

getdave committed:
Fix block gap inheritance for desktop container

- Add wp-block-navigation__desktop-container to gap: inherit rule
- Ensures navigation block continues to inherit block gap CSS after adding desktop container wrapper
- Overlay container excluded from gap inheritance as it controls its own spacing
- Fix indentation issues in padding rules

getdave committed:
Disable overlay menu for navigation blocks within overlay template parts

- Add disable_overlay_menu_for_nested_navigation_blocks() function to recursively find and modify navigation blocks
- Set overlayMenu attribute to 'never' for any navigation blocks found in overlay template parts
- Prevents nested overlays (inception problem) when navigation blocks are included in overlay content
- Applied to both theme file and database post code paths

getdave committed:
Add spacing to default close button when within custom overlay

- Add top and right spacing to wp-block-navigation__responsive-container-close when parent has has-custom-overlay class
- Use clamp value matching root padding pattern to position button away from overlay edges
- Only applies when default close button is rendered within a custom overlay template part

getdave committed:
Gate desktop container wrapper behind navigation overlay experiment

- Add is_overlay_experiment_enabled() helper to check experiment status
- Make desktop container wrapper conditional on experiment being enabled
- Maintain backward compatibility: without experiment, desktop navigation markup remains unchanged
- Prevents CSS/layout issues for users without the experiment enabled
- Add comprehensive comment explaining why this is gated

jeryj commented on packages/block-library/src/navigation/index.php:
```diff
@@ -272,6 +329,127 @@ private static function get_inner_blocks_from_fallback( $attributes ) {
+	 * @param array  $attributes                The block attributes.
+	 * @return WP_Block_List Returns the inner blocks for the overlay template part.
+	 */
+	private static function get_overlay_blocks_from_template_part( $overlay_template_part_id, $attributes ) {
```
I haven't worked with template parts much. This function is getting the blocks for a template by ID with sensible fallbacks, right? I'm surprised this isn't a util already.

jeryj commented on packages/block-library/src/navigation/index.php:
```diff
@@ -466,10 +644,12 @@ private static function get_responsive_container_markup( $attributes, $inner_blo
 		$responsive_container_classes = array(
 			'wp-block-navigation__responsive-container',
 			$is_hidden_by_default ? 'hidden-by-default' : '',
+			$has_custom_overlay ? 'has-custom-overlay' : '',
```
'has-custom-overlay' feels a little odd of a name since the overlay html is not within this code. Should we name it by intention, like 'disable-default-overlay'?

jeryj replied:
I'm rethinking this. I think the base navigation with default overlay (no custom overlay) should stay the same and not get any new classes. That way we're less likely to accidentally raise the specificity of things.

jeryj commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -499,6 +500,32 @@ button.wp-block-navigation-item__content {
+	// When menu is closed (not open), show desktop container and hide overlay container.
+	&:not(.is-menu-open) {
+		.wp-block-navigation__responsive-container-content {
+			.wp-block-navigation__desktop-container {
```
This menu might display on mobile devices too (no overlay, always show menu), so we should avoid "desktop" as the name. I'm trying to think of better names but struggling to find one I really like:

- default
- full
- text
- inline
- canvas
- page
- None? Just wp-block-navigation__container and have the overlay be the odd one out?

jeryj reviewed (commented):
I'm concerned about experimental overlay markup leaking out to the main canvas for people without the experiment on. Could the overlay specific stuff be more batched? Or have a default function that prefixes the functions and individual changes to make sure they don't get applied if the experiment is off?

jeryj commented:
I'm able to get the `has-custom-overlay` css class to show with the experiment off. Granted, you do have to turn it on first, but still, the states are leaking:

- Turn on the experiment
- Creat an overlay
- Assign it to a navigation
- Turn off the experiment
- View the frontpage menu
- Bug: Navigation has `has-custom-overlay` class
- Bug: Layout isn't right
- Bug: Close button is missing

<img width="589" height="757" alt="Screenshot 2025-12-12 at 10 19 41 AM" src="https://github.com/user-attachments/assets/0b7255c8-b695-4a46-a5d0-93c47cb9ee10" />

jeryj commented:
@getdave I rebased this with trunk to bring in the other recent overlay template changes

jeryj committed:
Consolate overlay code for easier gating

* sets default negative values for custom overlay experiment in one location
* consolates code to make it harder to accidentally bleed experiment code
* changes desktop class to inline so it's not screen size specific

jeryj committed:
Remove inline-container wrapper for custom overlays and simplify CSS

The wrapper could cause issues with how people target their css. Sibling + selectors could break if we add a new container. To avoid this, we use the existing markup and output the navigation custom overlay at the end of the block.

jeryj committed:
Remove custom color for focus outline

If the current color matches the background of the page (common) then the focus ring is not visible. Let it default to the color set by the page instead of overriding it.

jeryj committed:
Fix focus transfer and modal focus trap

On the default overlay, there are no extra focusable elements or duplicate menus. On a custom overlay, the existing menu is hidden and the custom overlay has additiona focusable elements. This was causing a problem for focus transfer and focus trap in the modal, as it was not checking for element visibilty when determining focusability.

jeryj committed:
Remove aria-hidden from custom overlay

This aria-hidden was not tied to the overlay state, so it was fixed as hidden. Also, this should be unnecssary, as display: none will already hide it from the accessibility tree.

scruffian commented on packages/block-library/src/navigation/index.php:
```diff
@@ -158,6 +170,51 @@ private static function get_markup_for_inner_block( $inner_block ) {
+	 * @param WP_Block_List $blocks The list of blocks to check.
+	 * @return bool Returns true if a navigation-overlay-close block is found.
+	 */
+	private static function has_navigation_overlay_close_block( $blocks ) {
```
I wonder if we should make this a general function for detecting whether a block contains inner blocks - we already do something like that in several places.

getdave replied:
Yes. I would suggest we do that in a followup as a code quality task

jeryj replied:
Did this in the current PR: https://github.com/WordPress/gutenberg/pull/73967/changes#diff-90cb5b64c95087ded3c7bb160e38189c8499b0f14f61beb884c887452449fa12R1262

scruffian commented on packages/block-library/src/navigation/index.php:
```diff
@@ -158,6 +170,51 @@ private static function get_markup_for_inner_block( $inner_block ) {
+				return true;
+			}
+
+			// Recursively check inner blocks, but skip navigation blocks.
```
Why skip navigation blocks?

jeryj replied:
A navigation block can't contain a navigation overlay close block, so we can safely skip navigation blocks.

getdave replied:
I agree the code comment isn't helpful though. Can we update that?

scruffian commented on packages/block-library/src/navigation/index.php:
```diff
@@ -465,11 +643,42 @@ private static function get_responsive_container_markup( $attributes, $inner_blo
+		$overlay_blocks_html = '';
+		$custom_overlay_markup = '';
+
+		if( $is_overlay_experiment_enabled ) {
```
```suggestion
		if ( $is_overlay_experiment_enabled ) {
```

scruffian commented on packages/block-library/src/navigation/index.php:
```diff
@@ -523,14 +732,34 @@ private static function get_responsive_container_markup( $attributes, $inner_blo
+		// Show default close button for all responsive navigation,
+		// unless custom overlay has its own close block.
+		if ( ! $has_custom_overlay_close_block ) {
+			$close_button_markup = sprintf(
```
What do you think about using an overlay close block here instead of the old close button markup?

scruffian replied:
This would also mean that we could keep the close button inside the overlay

jeryj replied:
That may be a safer option. Otherwise, we run into issues like this. (Right aligned menu with floating fixed position button). By auto-inserting the overlay close block, we make sure it's positioned outside of the content flow.

<img width="552" height="175" alt="close button overlapping navigation content" src="https://github.com/user-attachments/assets/1ee5953d-d3d7-4a32-806b-9defb100255b" />

jeryj replied:
Let's do it in a follow-up though. I think we should bring this in sooner.

jeryj commented:
> I'm wondering how users will set a custom background color for their overlays...

Setting a group block with 100VH as the root block works to have an overlay background color

jeryj committed:
Add generic block_tree_has_block_type function

We were using three separate functions that were doing the same thing - searching the block tree for the presence of a block. This refactors that into one generic function and implements it for each.

jeryj commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -523,21 +523,27 @@ button.wp-block-navigation-item__content {
-		padding-bottom: clamp(1rem, var(--wp--style--root--padding-bottom), 20rem);
-		padding-left: clamp(1rem, var(--wp--style--root--padding-left), 20rem);
+		// Don't apply default padding when custom overlay is present.
+		&:not(.disable-default-overlay) {
```
We should probably do this a different way. Adding the `:not` is going to raise the specificity of the default overlay padding, which people have likely overridden. It'll be safer to reset this at the custom overlay level.

jeryj commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -523,21 +523,27 @@ button.wp-block-navigation-item__content {
-			// Add padding above to accommodate close button.
+		// Add padding above to accommodate close button.
+		// Don't apply default padding when custom overlay is present.
+		&:not(.disable-default-overlay) .wp-block-navigation__responsive-container-content {
```
Same as above. This will raise the specificity. Let's try to find a way that doesn't raise it.

jeryj committed:
Fix php lint errors

scruffian commented on packages/block-library/src/navigation/index.php:
```diff
@@ -1020,26 +1250,54 @@ static function ( $block ) {
-function block_core_navigation_block_contains_core_navigation( $inner_blocks ) {
-	foreach ( $inner_blocks as $block ) {
-		if ( 'core/navigation' === $block->name ) {
+function block_core_navigation_block_tree_has_block_type( $blocks, $block_type, $skip_block_types = array() ) {
```
Could this have a more generic name?

jeryj replied:
I wanted to do 'block_tree_has_block_type' but the linter says I have to prefix with the navigation stuff. Maybe we can move it outside of the navigation block for wider use?

scruffian replied:
yeah i think we should, but maybe in a followup

scruffian committed:
fix tests

scruffian committed:
fix tests

getdave commented on phpunit/blocks/class-wp-navigation-block-renderer-test.php:
```diff
@@ -8,6 +8,15 @@
+			require_once __DIR__ . '/../../packages/block-library/src/navigation/index.php';
+		}
+	}
+
```
This is likely the source of test failures. This seems like an unusual pattern? Why is it needed? Can we simplify?

jeryj replied:
Good catch. I don't think it's necessary. AI written test that I missed it added setup.

jeryj replied:
I don't think we need the comment if nothing is happening there.

getdave committed:
fix: Update navigation block test to use prefixed functions

Fix test failure caused by directly requiring source file which bypassed
the build system. The test now correctly uses prefixed function names
from built files, following the build system function prefixing pattern.

Changes:
- Remove direct require of source file (functions loaded via lib/blocks.php)
- Update function calls to use gutenberg_ prefixed names
- Update @covers annotations to reference prefixed functions

See: docs/contributors/code/build-system-function-prefixing.md

jeryj committed:
Remove unnecessary setup from phpunit test

jeryj reviewed (approved):
I think this is safe to bring in now, and iterate on. I compared the HTML of a non-overlay menu with and without the experiment on and they matched except for the aria label. We should address this in a follow-up. 

aria labels get incremented within the overlay. So, if an aria-label used to say "Header Menu 2" it might now say "Header Menu 4" as it counts the menus within the overlay as well. We only want those to increment if they have an overlay (not set to overlay="never"). I looked into this for a bit today and I think that code needs improvement in general. I can apply a bandaid fix but it will be confusing based on the current code structure.

scruffian commented on packages/block-library/src/navigation/index.php:
```diff
@@ -467,9 +616,44 @@ private static function get_responsive_container_markup( $attributes, $inner_blo
 
 		$is_hidden_by_default = isset( $attributes['overlayMenu'] ) && 'always' === $attributes['overlayMenu'];
 
+		// Set-up variables for the custom overlay experiment.
```
This function is getting quite long. We might want to consider breaking this into smaller functions in a followup.

scruffian reviewed (approved):
This is looking good to me, let's bring it in.
